src/taskgraph/util/taskcluster.py

```diff
@@ -367,3 +368,22 @@ def list_task_group_incomplete_tasks(task_group_id):
             params = {"continuationToken": resp.get("continuationToken")}
         else:
             break
+
+
+def get_recursive_dependencies(
+    task_ids: List[str], add_given_tasks: bool = False, use_proxy: bool = False
+) -> Dict[str, str]:
+    """Returns a list of all the recursive dependencies of the task_ids provided as
+    a dictionary of a dictionary of label -> taskId."""
+    upstream_tasks: Dict[str, str] = {}
+
+    for task_id in task_ids:
+        task_def = get_task_definition(task_id, use_proxy)
+        if add_given_tasks:
+            upstream_tasks[task_def["metadata"]["name"]] = task_id
```

Comment:
I'm a bit fuzzy on why `add_given_tasks` would be useful. If the caller has the list of ids here already, can't they add these in themselves?

Reply:
I think it's necessary for this to work, unless we always add the given `task_ids` to the return value. (We have to after we've recursed, otherwise there's no base case, and the return value will always be empty.)

We could just always add them, but then the function becomes `get_ancestor_and_given_task_ids` or something, which I'm not sure is as useful.

The other way to do this is move the recursion out to a separate function:
```
def _get_deps(task_ids, use_proxy):
    upstream_tasks = {}
    for task_id in task_ids:
        task_def = get_task_definition(task_id, use_proxy)
        upstream_tasks[task_def["metadata"]["name"]] = task_id

        upstream_tasks.update(
            _get_deps(task_def["dependencies"], use_proxy)
        )

    return upstream_tasks

def get_recursive_dependencies(
    task_ids: List[str], use_proxy: bool = False
) -> Dict[str, str]:
    """Returns a list of all the recursive dependencies of the task_ids provided as
    a dictionary of a dictionary of label -> taskId."""
    upstream_tasks: Dict[str, str] = {}

    for task_id in task_ids:
        task_def = get_task_definition(task_id, use_proxy)

        upstream_tasks.update(
            _get_deps(task_def["dependencies"], use_proxy)
        )

    return upstream_tasks
```

Which certainly cleans up the interface. As you pointed out, there's really no good use case for `add_given_tasks` in the public interface, so maybe this is just better with no downsides...

Reply:
Thanks, I missed that this wasn't part of the public interface.. In that case I think this is fine. Maybe you could prefix it with an underscore and leave it out of the docstring (if you do end up documenting args there).

Reply:
I ended up switching to the above form, which I think reads nicer and improves the interface.